instrumentation.go

```diff
@@ -0,0 +1,280 @@
+package instana
+
+import (
+	"context"
+	"mime/multipart"
+	"net/http"
+	"net/textproto"
+	"net/url"
+
+	ot "github.com/opentracing/opentracing-go"
+	"github.com/opentracing/opentracing-go/ext"
+	otlog "github.com/opentracing/opentracing-go/log"
+)
+
+// TracingHandlerFunc is an HTTP middleware that captures the tracing data and ensures
+// trace context propagation via OpenTracing headers
+func TracingHandlerFunc(sensor *Sensor, name string, handler http.HandlerFunc) http.HandlerFunc {
+	return func(w http.ResponseWriter, req *http.Request) {
+		ctx := req.Context()
+
+		opts := []ot.StartSpanOption{
+			ext.SpanKindRPCServer,
+			ot.Tags{
+				string(ext.PeerHostname): req.Host,
+				string(ext.HTTPUrl):      req.URL.Path,
+				string(ext.HTTPMethod):   req.Method,
+			},
+		}
+
+		tracer := sensor.Tracer()
+		if ps, ok := SpanFromContext(req.Context()); ok {
+			tracer = ps.Tracer()
+			opts = append(opts, ot.ChildOf(ps.Context()))
+		}
+
+		wireContext, err := tracer.Extract(ot.HTTPHeaders, ot.HTTPHeadersCarrier(req.Header))
+		switch err {
+		case nil:
+			opts = append(opts, ext.RPCServerOption(wireContext))
+		case ot.ErrSpanContextNotFound:
+			log.debug("no span context provided with", req.Method, req.URL.Path)
+		case ot.ErrUnsupportedFormat:
+			log.info("unsupported span context format provided with", req.Method, req.URL.Path)
+		default:
+			log.warn("failed to extract span context from the request:", err)
+		}
+
+		span := tracer.StartSpan(name, opts...)
+		defer span.Finish()
+
+		defer func() {
+			// Be sure to capture any kind of panic / error
+			if err := recover(); err != nil {
+				if e, ok := err.(error); ok {
+					span.SetTag("message", e.Error())
```

Comment:
Do we really need the duplication?  The span has http tags so it will always be converted to an HTTPSpan in the backend - so only `http.error` needed right?

Reply:
I did a manual test and there seem be the same issue as we had with `rpc.error`: the error message from `http.error` is not being used as `message` and the error calls show up as an error without message.

Reply:
Here are examples of an erroneous call made with `http.error` only and with both `http.error` and `message` fields populated:
<img width="552" alt="Screenshot 2020-03-24 at 13 48 33" src="https://user-images.githubusercontent.com/203793/77427302-6f8a8780-6dd6-11ea-829b-3ce495001c65.png">